torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -662,6 +651,7 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
     // If not set before, get a dedicated stream for the device to run
     // FutureNCCL then callbacks.
     if (futureNCCLCallbackStreams_[deviceIndex] == nullptr) {
+      std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
This lock needs to be before the `if` statement. Otherwise two threads can concurrently enter this condition and unnecessarily call `getStreamFromPool` twice. To make this simpler, we can actually move this block of code after line 684 where we hold the lock. We can have an additional for loop there with the appropriate GPU guard set and only just set the NCCL callback streams.